File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/BaseTestCase.java
package com.clevertap.android.geofence;

import android.app.Application;
import android.os.Bundle;
import androidx.test.ext.junit.runners.AndroidJUnit4;
import org.junit.*;
import org.junit.runner.*;
import org.robolectric.annotation.Config;

@Config(sdk = 28,
        application = TestApplication.class
)
@RunWith(AndroidJUnit4.class)
public abstract class BaseTestCase {

    protected Application application;

    public static boolean areEqual(Bundle expected, Bundle actual) {
        if (expected == null) {
            return actual == null;
        }

        if (expected.size() != actual.size()) {
            return false;
        }

        for (String key : expected.keySet()) {
            if (!actual.containsKey(key)) {
                return false;
            }

            Object expectedValue = expected.get(key);
            Object actualValue = actual.get(key);

            if (expectedValue == null) {
                if (actualValue != null) {
                    return false;
                }

                continue;
            }

            if (expectedValue instanceof Bundle && actualValue instanceof Bundle) {
                if (!areEqual((Bundle) expectedValue, (Bundle) actualValue)) {
                    return false;
                }

                continue;
            }

            if (!expectedValue.equals(actualValue)) {
                return false;
            }
        }

        return true;
    }

    public static void assertBundlesEquals(String message, Bundle expected, Bundle actual) {
        if (!areEqual(expected, actual)) {
            Assert.fail(message + " <" + expected.toString() + "> is not equal to <" + actual.toString() + ">");
        }
    }

    public static void assertBundlesEquals(Bundle expected, Bundle actual) {
        assertBundlesEquals(null, expected, actual);
    }

    public Application getApplication() {
        return TestApplication.getApplication();
    }

    @Before
    public void setUp() throws Exception {
        application = TestApplication.getApplication();
    }

}

File: clevertap-geofence/src/test/java/com/clevertap/android/geofence/BaseTestCase.kt
package com.clevertap.android.geofence

import android.app.Application
import android.os.Bundle
import androidx.test.ext.junit.runners.AndroidJUnit4
import io.mockk.clearAllMocks
import org.junit.After
import org.junit.Assert
import org.junit.Before
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

@Config(sdk = [28], application = TestApplication::class)
@RunWith(AndroidJUnit4::class)
abstract class BaseTestCase {

    protected lateinit var application: Application

    @Before
    open fun setUp() {
        application = TestApplication.application
    }

    @After
    open fun cleanUp() {
        clearAllMocks()
    }

    companion object {
        fun areEqual(expected: Bundle?, actual: Bundle?): Boolean {
            if (expected == null) {
                return actual == null
            }

            if (actual == null) {
                return false
            }

            if (expected.size() != actual.size()) {
                return false
            }

            for (key in expected.keySet()) {
                if (!actual.containsKey(key)) {
                    return false
                }

                val expectedValue = expected.get(key)
                val actualValue = actual.get(key)

                if (expectedValue == null) {
                    if (actualValue != null) {
                        return false
                    }

                    continue
                }

                if (expectedValue is Bundle && actualValue is Bundle) {
                    if (!areEqual(expectedValue, actualValue)) {
                        return false
                    }

                    continue
                }

                if (expectedValue != actualValue) {
                    return false
                }
            }

            return true
        }

        fun assertBundlesEquals(message: String?, expected: Bundle, actual: Bundle) {
            if (!areEqual(expected, actual)) {
                Assert.fail("$message <$expected> is not equal to <$actual>")
            }
        }

        fun assertBundlesEquals(expected: Bundle, actual: Bundle) {
            assertBundlesEquals(null, expected, actual)
        }
    }
}
